File: Happy-Number.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(num) {
    function getNext(n) {
        let total = 0;
        while (n > 0) {
            let digit = n % 10;
            total += digit * digit;
            n = Math.floor(n / 10);
        }
        return total;
    }

    let seen = new Set();
    while (num !== 1 && !seen.has(num)) {
        seen.add(num);
        num = getNext(num);
    }

    return num === 1;
};

File: Happy-Number.java
public class Solution {
    public boolean isHappy(int num) {
        HashSet<Integer> seen = new HashSet<>();

        while (num != 1 && !seen.contains(num)) {
            seen.add(num);
            num = getNext(num);
        }

        return num == 1;
    }

    private int getNext(int n) {
        int total = 0;
        while (n > 0) {
            int digit = n % 10;
            total += digit * digit;
            n = n / 10;
        }
        return total;
    }
}